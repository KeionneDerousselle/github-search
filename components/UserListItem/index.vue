<template>
  <li
    v-bind="$attrs"
    class="user-list-item"
    v-on="$listeners">
    <img
      :src="user.avatar_url"
      :alt="`Profile picture for ${user.login}`"
      class="user-list-item__avatar">

    <div class="user-list-item__details">
      <span
        v-if="userDetails"
        class="user-list-item__details__name">{{ userDetails.name }}</span>
      <span class="user-list-item__details__login">{{ user.login }}</span>
    </div>

    <div
      v-if="userDetails"
      class="user-list-item__followers">
      {{ userDetails.followers }}
      <users-icon
        aria-hidden="true"
        class="user-list-item__followers__icon" />
      <span class="sr-only">Followers</span>
    </div>

    <a
      :href="user.html_url"
      target="_blank"
      rel="noopener"
      class="user-list-item__link">
      <link-icon
        class="user-list-item__link__icon"
        aria-hidden="true" />
      <span class="sr-only"> Opens link to {{ user.login }}'s GitHub Account in a new window.</span>
    </a>
  </li>
</template>
<script>
import LinkIcon from '@/components/Icons/Link'
import UsersIcon from '@/components/Icons/Users'

import { mapActions, mapGetters } from 'vuex'

export default {
  components: {
    LinkIcon,
    UsersIcon
  },

  props: {
    user: {
      type: Object,
      required: true
    }
  },

  data: () => ({
    userDetails: null
  }),

  computed: {
    ...mapGetters('users', ['userDetailsByUsername'])
  },

  beforeMount() {
    this.get({ username: this.user.login })
      .then(() => {
        const result = this.userDetailsByUsername(this.user.login)

        this.userDetails = result ? result.data : null
      })
  },

  methods: {
    ...mapActions('users', ['get'])
  }
}
</script>
<style lang="scss">
.user-list-item {
  @apply list-none flex items-center w-full border-b border-indigo-200 p-4;

  &:hover {
    @apply bg-indigo-50;
  }

  &__link {
    @apply ml-auto rounded-full border-transparent p-1 text-center flex justify-center items-center;
    @apply border;

    transition: border-color 0.3s ease-out;
  }

  &__link:hover,
  &__link:focus,
  &__link:active {
    @apply border-indigo-700 outline-none;

    box-shadow: none;
  }

  &__link__icon {
    @apply text-indigo-700;
    @apply w-4 h-4;
  }

  &__link__icon:hover,
  &__link__icon:focus {
    @apply outline-none border-0;
  }

  &__avatar {
    @apply inline-flex rounded-full mr-2.5;
    @apply w-10 h-10;
  }

  &__details {
    @apply inline-flex flex-col w-2/5;

    &__name {
      @apply text-gray-800 block;
      @apply sm:text-sm;
    }

    &__login {
      @apply text-gray-600 block;
      @apply text-xs;
      @apply sm:text-xs;
    }
  }

  &__followers {
    @apply text-indigo-700 inline-flex flex items-center ml-2.5 rounded-xl bg-indigo-100 px-1 py-0.5 text-sm font-bold border;
  }

  &__followers__icon {
    @apply w-4 h-4 text-indigo-700;
  }
}
</style>
