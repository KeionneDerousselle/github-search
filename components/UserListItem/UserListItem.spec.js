import UserListItem from '.'

describe('User List Item', () => {
  let wrapper
  let user
  let mockGetUserDetails
  let mockGetUserDetailsByUsername

  describe('when user details are not returned', () => {
    beforeAll(() => {
      user = {
        login: 'test_user',
        avatar_url: 'https://some.fake.com/avatar.png',
        html_url: 'https://www.some.external.link.com'
      }

      mockGetUserDetails = jest.fn().mockResolvedValue()
      mockGetUserDetailsByUsername = jest.fn().mockReturnValue()

      wrapper = shallowPreMocked(UserListItem, {
        store: {
          users: {
            actions: {
              get: mockGetUserDetails
            },
            getters: {
              userDetailsByUsername: jest.fn().mockReturnValue(mockGetUserDetailsByUsername)
            }
          }
        },
        propsData: {
          user
        }
      })
    })

    afterAll(() => {
      mockGetUserDetails.mockReset()
      wrapper.destroy()
    })

    it('should make a call to get the user\'s details', () => {
      expect(mockGetUserDetails).toHaveBeenCalledWith(expect.any(Object), expect.objectContaining({
        username: user.login
      }))
    })

    it('should render as expected', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should render the user\'s avatar', () => {
      const avatar = wrapper.get('.user-list-item__avatar')

      expect(avatar.element).toBeVisible()
      expect(avatar.attributes().src).toBe(user.avatar_url)
      expect(avatar.attributes().alt).toBe(`Profile picture for ${user.login}`)
    })

    it('should render the user\'s usernane', () => {
      const username = wrapper.get('.user-list-item__details__login')

      expect(username.element).toBeVisible()
      expect(username.text()).toBe(user.login)
    })

    it('should not render the user\'s followers', () => {
      expect(() => wrapper.get('.user-list-item__followers')).toThrow()
    })

    it('should not render the user\'s name', () => {
      expect(() => wrapper.get('.user-list-item__details__name')).toThrow()
    })

    it('should render an external link to the user\'s GitHub page', () => {
      const externalLink = wrapper.get('.user-list-item__link')

      expect(externalLink.attributes().href).toBe(user.html_url)
      expect(externalLink.attributes().target).toBe('_blank')
    })
  })

  describe('when user details are returned', () => {
    let userDetails

    beforeAll(() => {
      user = {
        login: 'test_user',
        avatar_url: 'https://some.fake.com/avatar.png',
        html_url: 'https://www.some.external.link.com'
      }

      userDetails = {
        name: 'Test User'
      }

      mockGetUserDetails = jest.fn().mockResolvedValue()
      mockGetUserDetailsByUsername = jest.fn().mockReturnValue({ data: userDetails })

      wrapper = shallowPreMocked(UserListItem, {
        store: {
          users: {
            actions: {
              get: mockGetUserDetails
            },
            getters: {
              userDetailsByUsername: jest.fn().mockReturnValue(mockGetUserDetailsByUsername)
            }
          }
        },
        propsData: {
          user
        }
      })
    })

    afterAll(() => {
      mockGetUserDetails.mockReset()
      wrapper.destroy()
    })

    it('should make a call to get the user\'s details', () => {
      expect(mockGetUserDetails).toHaveBeenCalledWith(expect.any(Object), expect.objectContaining({
        username: user.login
      }))
    })

    it('should render as expected', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should render the user\'s avatar', () => {
      const avatar = wrapper.get('.user-list-item__avatar')

      expect(avatar.element).toBeVisible()
      expect(avatar.attributes().src).toBe(user.avatar_url)
      expect(avatar.attributes().alt).toBe(`Profile picture for ${user.login}`)
    })

    it('should render the user\'s usernane', () => {
      const username = wrapper.get('.user-list-item__details__login')

      expect(username.element).toBeVisible()
      expect(username.text()).toBe(user.login)
    })

    it('should render the user\'s followers', () => {
      const followers = wrapper.get('.user-list-item__followers')

      expect(followers.element).toBeVisible()
    })

    it('should render the user\'s name', () => {
      const name = wrapper.get('.user-list-item__details__name')

      expect(name.element).toBeVisible()
      expect(name.text()).toBe(userDetails.name)
    })

    it('should render an external link to the user\'s GitHub page', () => {
      const externalLink = wrapper.get('.user-list-item__link')

      expect(externalLink.attributes().href).toBe(user.html_url)
      expect(externalLink.attributes().target).toBe('_blank')
    })
  })
})
