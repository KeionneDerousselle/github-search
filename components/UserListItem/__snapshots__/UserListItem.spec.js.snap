// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User List Item when user details are not returned should render as expected 1`] = `
<li class="user-list-item"><img src="https://some.fake.com/avatar.png" alt="Profile picture for test_user" class="user-list-item__avatar">
  <div class="user-list-item__details">
    <!----> <span class="user-list-item__details__login">test_user</span>
  </div>
  <!----> <a href="https://www.some.external.link.com" target="_blank" rel="noopener" class="user-list-item__link">
    <link-icon-stub aria-hidden="true" class="user-list-item__link__icon"></link-icon-stub> <span class="sr-only"> Opens link to test_user's GitHub Account in a new window.</span>
  </a>
</li>
`;

exports[`User List Item when user details are returned should render as expected 1`] = `
<li class="user-list-item"><img src="https://some.fake.com/avatar.png" alt="Profile picture for test_user" class="user-list-item__avatar">
  <div class="user-list-item__details"><span class="user-list-item__details__name">Test User</span> <span class="user-list-item__details__login">test_user</span></div>
  <div class="user-list-item__followers">

    <users-icon-stub aria-hidden="true" class="user-list-item__followers__icon"></users-icon-stub> <span class="sr-only">Followers</span>
  </div> <a href="https://www.some.external.link.com" target="_blank" rel="noopener" class="user-list-item__link">
    <link-icon-stub aria-hidden="true" class="user-list-item__link__icon"></link-icon-stub> <span class="sr-only"> Opens link to test_user's GitHub Account in a new window.</span>
  </a>
</li>
`;
